lib/src/file_util.rs

```diff
@@ -42,6 +42,25 @@ impl<T> IoResultExt<T> for io::Result<T> {
     }
 }
 
+/// Creates a directory and returns it's canonicalized path.
+///
+/// If the directory already exists, will only return it's canonicalized path
+/// without doing anything.
+///
+/// Returns the underlying error if the directory can't be created.
+/// The function will also fail if intermediate directories on the path do not
+/// already exist.
+pub fn create_or_reuse_dir(dirname: &Path) -> std::io::Result<PathBuf> {
+    match fs::create_dir(dirname) {
+        Ok(()) => {}
+        Err(_) if dirname.is_dir() => {}
+        Err(e) => return Err(e),
+    }
+
+    // if it fails, race condition?
+    dirname.canonicalize()
```

Comment:
Hmmm... You're right that it doesn't make sense to be in a file_util::create_or_reuse_dir().

I had initially put it into Workspace::create_root_dir() because it makes sense there as part of the 'protocol' for creating a new workspace directory is to canonicalize that directory, else, each potential caller always has to remember to canonicalized the returned directory.

I'm thinking I should just move this wholesale back to Workspace::create_root_dir(), but let me know if you feel strongly about this.

Reply:
> Hmmm... You're right that it doesn't make sense to be in a file_util::create_or_reuse_dir().
> 
> I had initially put it into Workspace::create_root_dir() because it makes sense there as part of the 'protocol' for creating a new workspace directory is to canonicalize that directory, else, each potential caller always has to remember to canonicalized the returned directory.

I tend to agree that the path canonicalization in Workspace and Repo are messy, but I don't think the public functions of `Workspace` should require callers to canonicalize the workspace root.

If we can move `create_dir()` to `Workspace::init_*()`, I think that would be the best. If we can't, I wouldn't change the expectation of `Workspace::init_*()` functions.

Reply:
I didn't find any better way to do this, so I'm staying with your suggestion of .and_then().